src/Workspaces/CoreTestUtilities/ExportProviderCache.cs

```diff
@@ -103,45 +139,67 @@ public static ComposableCatalog WithPart(this ComposableCatalog catalog, Type t)
+            }
+            else if (s_remoteHostExportProviderFactory != null && catalog == s_remoteHostCatalog)
+            {
+                return s_remoteHostExportProviderFactory;
             }
 
-            var expectedCatalog = Interlocked.CompareExchange(ref _expectedCatalog, catalog, null) ?? catalog;
-            if (expectedCatalog != catalog)
+            var configuration = CompositionConfiguration.Create(catalog.WithCompositionService());
+            var runtimeComposition = RuntimeComposition.CreateRuntimeComposition(configuration);
+            var exportProviderFactory = runtimeComposition.CreateExportProviderFactory();
+            return new SingleExportProviderFactory(catalog, exportProviderFactory);
+        }
+
+        private class SingleExportProviderFactory : IExportProviderFactory
+        {
+            private readonly ComposableCatalog _catalog;
+            private readonly IExportProviderFactory _exportProviderFactory;
+
+            public SingleExportProviderFactory(ComposableCatalog catalog, IExportProviderFactory exportProviderFactory)
             {
-                throw new InvalidOperationException($"Only one {nameof(ExportProvider)} can be created for a single test.");
+                _catalog = catalog;
+                _exportProviderFactory = exportProviderFactory;
             }
 
-            var expected = _expectedProviderForCatalog;
-            if (expected == null)
+            public ExportProvider CreateExportProvider()
             {
-                if (!_exportProviderFactoryCache.TryGetValue(catalog, out var exportProviderFactory))
+                if (!_enabled)
                 {
-                    var configuration = CompositionConfiguration.Create(catalog.WithCompositionService());
-                    var runtimeComposition = RuntimeComposition.CreateRuntimeComposition(configuration);
-                    exportProviderFactory = runtimeComposition.CreateExportProviderFactory();
-                    if (cacheExportProviderFactory)
-                    {
-                        _exportProviderFactoryCache.Add(catalog, exportProviderFactory);
-                    }
+                    throw new InvalidOperationException($"{nameof(ExportProviderCache)} may only be used from tests marked with {nameof(UseExportProviderAttribute)}");
                 }
 
-                expected = exportProviderFactory.CreateExportProvider();
-                expected = Interlocked.CompareExchange(ref _expectedProviderForCatalog, expected, null) ?? expected;
-                Interlocked.CompareExchange(ref _currentExportProvider, expected, null);
-            }
+                var expectedCatalog = Interlocked.CompareExchange(ref _expectedCatalog, _catalog, null) ?? _catalog;
+                if (expectedCatalog != _catalog)
+                {
+                    throw new InvalidOperationException($"Only one {nameof(ExportProvider)} can be created for a single test.");
```

Comment:
I admit I'm still not getting the one-export-provider restriction. If it's for cleanup, why can't we just clean all of the ones that were made up? It seems like it'd simplify the code.

I can imagine you had cases through this work where we were creating one that was previously a static and by re-creating multiple instances you caused a bug because state that was a singleton in a test now isn't. But it seems that was a point-in-time offense, where now just allowing multiple to be created and cleaning up all of them would  be easier code to maintain going forward.

Put another way: this rule might have been great for your work to do this, but seems like it needs justification (and commenting) if it's going to stay.

Reply:
➡️ Documenting code at the site where the exception is thrown